packages/babel-parser/src/parser/expression.js

```diff
@@ -2496,17 +2496,30 @@ export default class ExpressionParser extends LValParser {
     );
 
     this.next();
-    if (
-      this.match(tt.semi) ||
-      (!this.match(tt.star) && !this.state.type.startsExpr) ||
-      this.hasPrecedingLineBreak()
-    ) {
-      node.delegate = false;
-      node.argument = null;
-    } else {
-      node.delegate = this.eat(tt.star);
-      node.argument = this.parseMaybeAssign();
+    let delegating = false;
+    let argument = null;
+    if (!this.hasPrecedingLineBreak()) {
+      delegating = this.eat(tt.star);
+      switch (this.state.type) {
+        case tt.semi:
+        case tt.eof:
+        case tt.braceR:
+        case tt.parenR:
+        case tt.bracketR:
+        case tt.braceBarR:
+        case tt.colon:
+        case tt.comma:
```

Comment:
The original approach checks `tt.semi` and then excludes tokens with `type.startsExpr`. I think the new approach easier to reason about since it is an allowlist. This approach is taken from V8:

https://source.chromium.org/chromium/chromium/src/+/main:v8/src/parsing/parser-base.h;l=2979;drc=d17745f350d4956a07cb1113ee19e9cbc4be699f;bpv=1;bpt=1

However, V8 has `tt.colon` here, which I think it is redundant because it seems to me that `:` never follow an AssignmentExpression. The spec has

```
[+In] RelationalExpression[?In, ?Yield, ?Await] in ShiftExpression
for ( LeftHandSideExpression in Expression ) Statement
for ( var ForBinding in Expression ) Statement
for ( ForDeclaration in Expression ) Statement
```

None of the productions before in can parsed down to an argument-less YieldExpression.

Reply:
Maybe it's because of this?
```js
function* fn() {
  a ? yield : 2
}
```

Reply:
Oh I meant for `tt._in`. 🤦